src/main/java/htsjdk/utils/ClassFinder.java

```diff
@@ -0,0 +1,192 @@
+    }
+
+    public ClassFinder(final ClassLoader loader) {
+        this.loader = loader;
+    }
+
+    public ClassFinder(final File jarFile) throws IOException {
+        // The class loader must have the context in order to load dependent classes when loading a class,
+        // but the jarPath is remembered so that the iteration over the classpath skips anything other than
+        // the jarPath.
+        jarPath = jarFile.getCanonicalPath();
+        final URL[] urls = {new File(jarPath).toURI().toURL()};
+        loader = new URLClassLoader(urls, Thread.currentThread().getContextClassLoader());
+    }
+
+    /** Convert a filename to a class name by removing '.class' and converting '/'s to '.'s. */
+    public String toClassName(final String filename) {
+        return filename.substring(0, filename.lastIndexOf(".class"))
+                .replace('/', '.').replace('\\', '.');
+    }
+
+    /**
+     * Scans the classpath for classes within the specified package and sub-packages that
+     * extend the parentType. This method can be called repeatedly
+     * with different packages. Classes are accumulated internally and
+     * can be accessed by calling {@link #getClasses()}.
+     */
+    public void find(String packageName, final Class<?> parentType) {
+        this.parentType = parentType;
+        packageName = packageName.replace('.', '/');
+        final Enumeration<URL> urls;
+
+        try {
+            urls = loader.getResources(packageName);
+        }
+        catch (IOException ioe) {
+            log.warn("Could not read package: " + packageName, ioe);
+            return;
+        }
+
+        while (urls.hasMoreElements()) {
+            try {
+                String urlPath = urls.nextElement().getFile();
+                // convert URL to URI
+                // http://bugs.java.com/bugdatabase/view_bug.do?bug_id=4466485
+                // using URLDecode does not work if urlPath has a '+' character
+                try {
+                    URI uri = new URI(urlPath);
+                    urlPath = uri.getPath();
+                } catch (URISyntaxException e) {
+                    log.warn("Cannot convert to URI the " + urlPath + " URL");
+                }
+                if (urlPath.indexOf('!') > 0) {
+                    urlPath = urlPath.substring(0, urlPath.indexOf('!'));
+                }
+                if (jarPath != null && !jarPath.equals(urlPath)) {
+                    continue;
+                }
+
+                //Log.info("Looking for classes in location: " + urlPath);
+                final File file = new File(urlPath);
+                if ( file.isDirectory() ) {
+                    scanDir(file, packageName);
+                }
+                else {
+                    scanJar(file, packageName);
+                }
+            }
+            catch (IOException ioe) {
+                log.warn("could not read entries", ioe);
```

Comment:
Would we be better off removing these catch and log statements?  It seems like they could potentially hide critical errors.  We have them in gatk as well, but I'm not sure they're the right thing to do.

Reply:
I copied it verbatim from Picard (which is probably how it got into GATK...) so I would argue that it's out of scope for this PR....do you agree?

Reply:
Agreed, more of a thought for the future.